### PR TITLE
Add stomp protocol implementation

### PR DESCRIPTION
**Pull request by kazeborja**
This PR makes a bit more complete the STOMP protocol file of the implementation.

Things that I think need to be improved:
- Handling publish message error: publish should be retried with a new connection (to hopefully reach a different message broker), in case publication keeps failling message should be discarded? archived?
- Handling reconnection: when an error occurs it should try to reconnect to a different message broker (maybe here I can resolve the list of addresses behing the alias for CERN MB and keep them in memory to loop over them?) and retry until no more addresses are available
- In case the connection can't be stablished, Shoveler should still take messages in and put them in queue, then resume when a connection can be stablished.

### CODE REVIEW

**Review by djw8605 — CHANGES_REQUESTED** [review]
I don't think you need a separate goroutine for the handleReconnect.  Instead, I think you should attempt a reconnect only when the publish fails.  The publish should be in a while loop, with a retry delay.  So the publish function would look something like:
```
for {
    err := session.conn.Send(
		session.topic,
		"text/plain",
		msg)
    if err != nil {
        log.Errorln("Failed to send message, will retry:", err.Error)
reconnectLoop:
        for {
            reconnectError := handleReconnect()
            if reconnectError == nil {
                break reconnectLoop
            } else {
                log.Errorln("Failed to reconnect:", reconnectError.Error)
            }
        }
    }
}
```
Or something like that.  The reconnect loop could also be in the handleReconnect function.

File: stomp.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+// handleReconnect reconnects to the stomp server
+func (session *StompSession) handleReconnect() {
+	// Close the current session
+	session.conn.Disconnect()
+
+	// Start a new session
+	conn, err := stomp.Dial("tcp", session.stompUrl.String(),
+		stomp.ConnOpt.Login(session.username, session.password))
+	if err != nil {
+		log.Errorln("Failed to connect. Retrying:", err.Error)
```
[the remark this answers is not in the document]
Comment (kazeborja) [reply stomp.go]: Indeed, I think when I started writing this on Friday I left it like (it was done) but it wasn't :P

File: stomp.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+// publish will send the message to the stomp message bus
+// It will also handle any error in sending by calling handleReconnect
+func (session *StompSession) publish(msg []byte) error {
+	err := session.conn.Send(
+		session.topic,
+		"text/plain",
+		msg)
+
+	return err
```
[the remark this answers is not in the document]
Comment (kazeborja) [reply stomp.go]: I've added the change here, so it tries to reconnect when publish fails and keeps trying to publish the message until achieved

File: stomp.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+// It will also handle any error in sending by calling handleReconnect
+func (session *StompSession) publish(msg []byte) {
+sendMessageLoop:
+	for {
+		err := session.conn.Send(
+			session.topic,
+			"text/plain",
+			msg)
+
+		if err != nil {
```
Comment (djw8605) [inline stomp.go:80]: I think it's worth it to log the err from Send.

File: stomp.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		reconnectLoop:
+			for {
+				reconnectError := session.handleReconnect()
+				if reconnectError == nil {
+					break reconnectLoop
+				} else {
+					log.Errorln("Failed to reconnect:", reconnectError.Error())
```
Comment (djw8605) [inline stomp.go:87]: Add a delay here so it's not rapidly failing.  I think after the log, just add:

```
<-time.After(reconnectDelay)
```

And set reconnectDelay to something.

Comment (kazeborja) [reply stomp.go]: Since the constant was defined in amqp I've reused it and extracted it into a common const file, let me know what you think

**Comment by djw8605** [comment]
I think just fix the linting issues and we're good to go.

File: stomp.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		err := session.conn.Send(
+			session.topic,
+			"text/plain",
+			msg)
+
+		if err != nil {
+			log.Errorln("Failed to publish message:", err)
+			session.handleReconnect()
+		} else {
+			break sendMessageLoop
```
Comment (djw8605) [inline stomp.go:94]: I suppose this doesn't need to be a labeled break, there is only 1 loop.  Also, it could probably just be a return.

Comment (kazeborja) [reply stomp.go:94]: On that I'll take any recommendation as I'm not quite familiar with go standards, the concept of labeled loops/breaks semeed easier to read but that's about it

Comment (djw8605) [reply stomp.go:94]: You are probably as familiar with go standards as me.  Whatever is most readable, I agree though that labeled loops are nice.

So if you are happy with this, I think we're good.

File: stomp.go [hunk 6]
```diff
@@ -14,7 +14,7 @@ func StartStomp(config *Config, queue *ConfirmationQueue) {
 	stompUser := config.StompUser
 	stompPassword := config.StompPassword
 	stompUrl := config.StompURL
-	stompTopic := config.StompTopic
+	stompTopic := "/topic/" + config.StompTopic
```
Comment (djw8605) [inline stomp.go:17]: Maybe, for sanity, you could check of the `config.StompTopic` starts with `/topic/`, and prepend it if it does not.

**Pull request merged**